renderdoc/driver/vulkan/vk_pixelhistory.cpp

```diff
@@ -1410,6 +1441,671 @@ struct TestsFailedCallback : public VulkanPixelHistoryCallback
+  void PreEndCommandBuffer(VkCommandBuffer cmd) {}
+  void AliasEvent(uint32_t primary, uint32_t alias) {}
+  uint32_t GetEventOffset(uint32_t eid)
+  {
+    auto it = m_EventIndices.find(eid);
+    RDCASSERT(it != m_EventIndices.end());
+    return it->second;
+  }
+
+private:
+  // For each event, specifies where the occlusion query results start.
+  std::map<uint32_t, uint32_t> m_EventIndices;
+  // Number of fragments for each event.
+  std::map<uint32_t, uint32_t> m_EventFragments;
+  // Original image for which pixel history was requested.
+  VkImage m_Image;
+  VkFormat m_SrcFormat;
+  VkBuffer m_DstBuffer;
+  // Number of fragments processed so far.
+  uint32_t fragsProcessed = 0;
+  // These were created with VK_FORMAT_R32G32B32A32_SFLOAT format.
+  // The image is cleared between each replay for a fragment, and is
+  // used to get shader output value and primitive ID.
+  VkImageView m_PerFragmentImageView;
+  VkImage m_PerFragmentImage;
+
+  rdcarray<VkPipeline> m_PipesToDestroy;
+  rdcarray<VkRenderPass> m_RpsToDestroy;
+  rdcarray<VkFramebuffer> m_FbsToDestroy;
+};
+
+// Callback used to determine the shader discard status for each fragment, where
+// an event has multiple fragments with some being discarded in a fragment shader.
+struct VulkanPixelHistoryDiscardedFragmentsCallback : VulkanPixelHistoryCallback
+{
+  // Key is event ID and value is a list of primitive IDs
+  std::map<uint32_t, rdcarray<int32_t> > m_Events;
+  VulkanPixelHistoryDiscardedFragmentsCallback(WrappedVulkan *vk,
+                                               PixelHistoryShaderCache *shaderCache, uint32_t x,
+                                               uint32_t y, uint32_t sampleMask,
+                                               std::map<uint32_t, rdcarray<int32_t> > events,
+                                               VkQueryPool occlusionPool)
+      : VulkanPixelHistoryCallback(vk, shaderCache, x, y, sampleMask, occlusionPool),
+        m_Events(events)
+  {
+  }
+
+  ~VulkanPixelHistoryDiscardedFragmentsCallback()
+  {
+    for(const VkPipeline &pipe : m_PipesToDestroy)
+      m_pDriver->vkDestroyPipeline(m_pDriver->GetDev(), pipe, NULL);
+  }
+
+  void PreDraw(uint32_t eid, VkCommandBuffer cmd)
+  {
+    if(m_Events.find(eid) == m_Events.end())
+      return;
+
+    const rdcarray<int32_t> primIds = m_Events[eid];
+
+    VulkanRenderState prevState = m_pDriver->GetCmdRenderState();
+    VulkanRenderState &state = m_pDriver->GetCmdRenderState();
+    const VulkanCreationInfo::Pipeline &p =
+        m_pDriver->GetDebugManager()->GetPipelineInfo(state.graphics.pipeline);
+    // Create a pipeline with a scissor and colorWriteMask = 0, and disable all tests.
+    VkPipeline newPipe =
+        CreatePipeline(state.graphics.pipeline, eid, p.dynamicStates[VkDynamicScissor]);
+    state.graphics.pipeline = GetResID(newPipe);
+    state.BindPipeline(m_pDriver, cmd, VulkanRenderState::PipelineBinding::BindGraphics, false);
```

Comment:
`PipelineBinding` is a normal enum so the extra scope here is unnecessary.

Reply:
Fixed